Add rules to check security requirements

mikekistler opened the pull request:
This PR adds three new rules to check various aspects of the global and operation-level `security` property.

- az-operation-security checks that every operation has a security requirement (may be defined globally).
- az-security-min-length checks that a security property always has at least one entry
- az-security-requirements checks that each property in a security requirement object references a defined security scheme and that oauth2 schemes have a non-empty scopes array with scopes that are all defined.

Docs and tests are included.

I also re-enabled oas2-unused-definition. I suppose I initially turned this off because it could generate a bunch of messages for some services, but I think it's better to see these and consciously ignore them rather than not see them at all.

MushMal commented on test/operation-security.test.js:
```diff
@@ -0,0 +1,125 @@
+  });
+});
+
+test('az-operation-security should find operations without security', () => {
```
Should the name of the test specify the document type that's it's targeting? As this has the same name as the above test.

mikekistler replied:
Agreed.  I wish jest allowed grouping of tests so that I could have a separate group for oas3 with the same descriptions, but AFAIK it does not.  I will fix.

MushMal commented on test/operation-security.test.js:
```diff
@@ -0,0 +1,125 @@
+  });
+});
+
+test('az-operation-security should find no errors', () => {
```
same about the name of the test here and the last test

MushMal commented on test/security-requirements.test.js:
```diff
@@ -0,0 +1,311 @@
+  };
+  return linter.run(oasDoc).then((results) => {
+    expect(results.length).toBe(1);
+    expect(results[0].path.join('.')).toBe('security.0.AADToken');
```
For my understanding... Where do these values come from?

mikekistler replied:
It is just the JSON Path to the element that has the error.  Here it is the "AADToken" property of the 0th element of the "security" property at the top level of the document.

MushMal reviewed (commented):
Minor comments. Most are questions for my understanding.

mikekistler commented on functions/security-requirements.js:
```diff
@@ -0,0 +1,73 @@
+        const scheme = schemes[key];
+        // oas2-schema requires scheme to be an object.
+        // Checking here just to avoid runtime errors.
+        if (!isObject(scheme)) { return; }
```
I think LN40 will be tested by [the test for malformed API doc](https://github.com/Azure/azure-api-style-guide/pull/82/files#diff-2e4fcd01a70a3982c625cb45fff6916518662e9b8b887d34a945748eeaf47eceR249), because the "oauth2" security scheme is not an object.

mikekistler commented on functions/security-requirements.js:
```diff
@@ -0,0 +1,73 @@
+              });
+            }
+          });
+        } else if (scopes.length > 0) {
```
Correct.

mikekistler commented:
@MushMal I pushed a new commit that think addresses your comments.  I also had to rebase to fix a git conflict, but you should be able to just look at the last commit to see my changes.
